### PR TITLE
Antlr4

### PR DESCRIPTION
**Pull request by ghost**
Parse expr with ANTLR4 instead of Parboiled, because ANTLR4 works with Kotlin 1.5 and on Java 17.

### CODE REVIEW

File: src/main/antlr4/Expr.g4 [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+date : 'd\'' (SQESC|':'|.)*? '\'';
+time : 't\'' (SQESC|':'|.)*? '\'';
+datetime: 'dt\'' (SQESC|':'|.)*? '\'';
+fragment SQESC: '\\\'' | '\\\\';
+
+varName: VAR_NAME;
+VAR_NAME: [a-zA-Z_][a-zA-Z0-9_]*;
+
+list : '[' ( value ( ',' value )+ ','? )? ']';
```
Comment (BenSmith-MFJ) [inline src/main/antlr4/Expr.g4:61]: This is not identical to the expression in the readme, there's a `+` v `*`. Got me thinking about some more tests. I added them, take them or leave them as you like. They passed, for what it's worth.

https://github.com/measuresforjustice/expr/pull/11

**Review by BenSmith-MFJ — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -31,6 +33,8 @@ _Expr_ is deployed to [Maven Central](https://repo1.maven.org/maven2/io/mfj/expr
 * List:             '[' ( Value ( ',' Value )* ','? )? ']'
 ```
 
+[ANTLR4 Grammar](src/main/resources/io/mfj/expr/Expr.g4)
```
[the remark this answers is not in the document]
Comment (stephen-byrne) [reply README.md]: fixed in 03ca6d9

File: src/main/antlr4/Expr.g4 (hunk 1 above)
Comment (stephen-byrne) [reply src/main/antlr4/Expr.g4:67]: The last `?` makes it work. I merged in your tests at 5077350. They are definitely good things to test, thank you.

File: src/main/antlr4/Expr.g4 [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+fragment OR : [oO][rR];
+
+OPERATOR : '!=' | '<>' | '>=' | '<=' | '>' | '<' | '=~' | '=' | IN | CONTAINS ;
+
+fragment IN : '!'?[iI][nN];
+fragment CONTAINS: '!'?[cC][oO][nN][tT][aA][iI][nN][sS];
+
+literalValue : nul | number | string | bool | regex | date | time | datetime;
+
+nul: 'null';
```
Comment (whbboyd) [inline src/main/antlr4/Expr.g4:36]: The inconsistent case-sensitivity (e.g. if I'm reading correctly, `null` is case-sensitive while `not` is not) is surprising to me. Is that matching previous behavior? IIRC, the parboiled parser was completely case-insensitive for language keywords; but even if it wasn't, it seems like we probably should be either completely case-sensitive or completely case-insensitive, not a mix depending on the specific keyword.

**Pull request merged**